### PR TITLE
[Mailer] Add support for allowing some users even if `recipients` is defined in `EnvelopeListener`

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

I'm migrate an application from SwiftMailer to symfony/mailer, and this options was used.

---

depends on #54292

### CODE REVIEW

**Review by welcoMattic — APPROVED** [review]

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 1]
```diff
@@ -30,17 +31,25 @@ class EnvelopeListener implements EventSubscriberInterface
...
+     * @var string[]
+     */
+    private array $allowedRecipients;
+
     /**
      * @param array<Address|string> $recipients
+     * @param string[]              $allowedRecipients An array of regex to match the allowed recipients
      */
-    public function __construct(Address|string|null $sender = null, ?array $recipients = null)
+    public function __construct(Address|string|null $sender = null, ?array $recipients = null, ?array $allowedRecipients = null)
```
Comment (stloyd) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: You pass `[]` anyway:
```suggestion
    public function __construct(Address|string|null $sender = null, ?array $recipients = null, array $allowedRecipients = [])
```

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 2]
```diff
@@ -30,17 +31,25 @@ class EnvelopeListener implements EventSubscriberInterface
...
+     * @var string[]
+     */
+    private array $allowedRecipients;
+
     /**
      * @param array<Address|string> $recipients
+     * @param string[]              $allowedRecipients An array of regex to match the allowed recipients
      */
-    public function __construct(Address|string|null $sender = null, ?array $recipients = null)
+    public function __construct(Address|string|null $sender = null, ?array $recipients = null, array $allowedRecipients = [])
```
Comment (OskarStark) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: Lets use a promoted property

Comment (lyrixx) [reply src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: fixed, thanks

**Comment by nicolas-grekas** [comment]
I don't understand what's this for, can you please give me a few examples ? The code is fairly intricated for only two options and I'm wondering how this logic is sensible for real world use cases. :pray:

**Comment by lyrixx** [comment]
Sure, [here is the documentation ](https://symfony.com/doc/4.x/email.html#sending-to-a-specified-address-but-with-exceptions)

And our real world example is for the preproduction:

```yaml
mailer_recipients: [ 'REDACTED@jolicode.com' ]
mailer_recipients_allowed: [ '/@jolicode\.com$/', '/^foobar@example\.com$/', '/^baz@example\.com$/' ]
```

**TL;DR:**

in preprod, we want that all mails to be sent to `'REDACTED@jolicode.com'` except the ones targeting our agency, or some specials client mail addresses

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -2126,6 +2126,15 @@ private function addMailerSection(ArrayNodeDefinition $rootNode, callable $enabl
                                     ->end()
                                     ->prototype('scalar')->end()
                                 ->end()
+                                ->arrayNode('recipients_allowed')
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: This is missing the `fixXmlConfig` on the parent node which is required to properly support XML config files where we have multiple tags named in a singular.

And this misses the updated of the XSD file as well.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -2126,6 +2126,15 @@ private function addMailerSection(ArrayNodeDefinition $rootNode, callable $enabl
...
+                                    ->info('A list of regular expressions that allow recipients when "recipients" option is defined.')
+                                    ->performNoDeepMerging()
+                                    ->beforeNormalization()
+                                    ->ifArray()
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: should be indented as this is part of the `beforeNormalization` configuration.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -2126,6 +2126,15 @@ private function addMailerSection(ArrayNodeDefinition $rootNode, callable $enabl
...
+                                        ->then(fn ($v) => array_filter(array_values($v)))
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:2137]: shouldn't this do `array_values` after filtering if you want to return a list ? Otherwise, I don't see the benefit.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: btw, I would name it `allowed_recipients` as adjectives come before the noun in English.

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 6]
```diff
@@ -32,15 +33,21 @@ class EnvelopeListener implements EventSubscriberInterface
...
+        private array $allowedRecipients = [],
+    ) {
         if (null !== $sender) {
             $this->sender = Address::create($sender);
         }
         if (null !== $recipients) {
             $this->recipients = Address::createArray($recipients);
         }
+
+        $this->allowedRecipients = $allowedRecipients;
```
Comment (stof) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: this is useless as you made it a promoted property. It is already assigned

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 5 above)
Comment (lyrixx) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:2137]: Actually, I copied what is done on `->arrayNode('recipients')`

**Comment by lyrixx** [comment]
I added some tests for the DIC part. But the configuration was broken in XML. I need #54292 first to be merged

**Review by stof — APPROVED** [review]

**Comment by lyrixx** [comment]
I have rebased, now the PR is to merge

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 7]
```diff
@@ -57,7 +62,24 @@ public function onMessage(MessageEvent $event): void
...
+                            $recipients[] = $recipient;
+                            continue 2;
+                        }
+                    }
+                    $addDefaults = true;
+                }
+                if ($addDefaults) {
+                    $recipients = array_merge($recipients, $this->recipients);
+                }
+            }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php:83]: Looking at this logic: if one recipient isn't allowed, then we send to all $recipents, but if all recipients are allowed, then we don't send to $recipents. Shouldn't we always send to $recipents and in addition also send to allowed recipients? (and possibly dedup?)

Comment (lyrixx) [reply src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php:83]: okay, done ✅

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 8]
```diff
@@ -57,7 +62,27 @@ public function onMessage(MessageEvent $event): void
         }
 
         if ($this->recipients) {
-            $event->getEnvelope()->setRecipients($this->recipients);
+            $recipients = $this->recipients;
+            if ($this->allowedRecipients) {
+                foreach ($event->getEnvelope()->getRecipients() as $recipient) {
+                    foreach ($this->allowedRecipients as $allowedRecipient) {
+                        if (!preg_match($allowedRecipient, $recipient->getAddress())) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: I'm wondering if we shouldn't anchor regexps, so that the config is less error prone and easier to write:
```suggestion
                        if (!preg_match('{'.$allowedRecipient.'}A', $recipient->getAddress())) {
```

this would allow listing `'.*@example.com'` to allow all emails from this domain (instead of `'/@example.com$/'` at the moment)

Comment (lyrixx) [reply src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]: pr updated

File: src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php [hunk 9]
```diff
@@ -57,7 +62,27 @@ public function onMessage(MessageEvent $event): void
         }
 
         if ($this->recipients) {
-            $event->getEnvelope()->setRecipients($this->recipients);
+            $recipients = $this->recipients;
+            if ($this->allowedRecipients) {
+                foreach ($event->getEnvelope()->getRecipients() as $recipient) {
+                    foreach ($this->allowedRecipients as $allowedRecipient) {
+                        if (!preg_match('{'.$allowedRecipient.'}A', $recipient->getAddress())) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/EventListener/EnvelopeListener.php]:
```suggestion
                        if (!preg_match('{\A'.$allowedRecipient.'\z}', $recipient->getAddress())) {
```

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/mailer_with_dsn.php [hunk 10]
```diff
@@ -13,6 +13,7 @@
             'envelope' => [
                 'sender' => 'sender@example.org',
                 'recipients' => ['redirected@example.org'],
+                'allowed_recipients' => ['/^foobar@example.org$/'],
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/php/mailer_with_dsn.php]: same in all examples:
```suggestion
                'allowed_recipients' => ['foobar@example\.org'],
```

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/mailer_with_transports.xml [hunk 11]
```diff
@@ -16,6 +16,8 @@
                 <framework:sender>sender@example.org</framework:sender>
                 <framework:recipient>redirected@example.org</framework:recipient>
                 <framework:recipient>redirected1@example.org</framework:recipient>
+                <framework:allowed-recipient>/^foobar@example.org$/</framework:allowed-recipient>
+                <framework:allowed-recipient>/@example.com$/</framework:allowed-recipient>
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/mailer_with_transports.xml]:
```suggestion
                <framework:allowed-recipient>.*@example\.com</framework:allowed-recipient>
```

File: src/Symfony/Component/Mailer/Tests/EventListener/EnvelopeListenerTest.php [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+        $listener->onMessage($event);
+
+        $recipients = array_map(fn (Address $a): string => $a->getAddress(), $event->getEnvelope()->getRecipients());
+        $this->assertSame($expected, $recipients);
+    }
+
+    public static function provideRecipientsTests(): iterable
+    {
+        yield [['r1@example.com', 'r2@symfony.com'], null, []];
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Mailer/Tests/EventListener/EnvelopeListenerTest.php:41]: we need to escape all dots:
```suggestion
        yield [['r1@example\.com', 'r2@symfony\.com'], null, []];
```

**Comment by lyrixx** [comment]
@nicolas-grekas I have addressed your comments

**Review by fabpot — APPROVED** [review]
@lyrixx We should probably add an entry in the framework bundle changelog as well like:

```
 * Add the `allowed_recipients` option for mailer to allow some users to receive emails even if `recipients` is defined.
```

**Comment by lyrixx** [comment]
@fabpot thanks. Added (+ rebased)

**Comment by fabpot** [comment]
Thank you @lyrixx.
